docs/src/best_practices.md

```diff
@@ -12,4 +12,32 @@ From the perspective of software, we can break the neatly into two components. T
 
-## Testing
\ No newline at end of file
+## Testing
+
+Here, we introduce the testing methodology for software, based on the `Test.jl` package in Julia's standard library. As software gets developed, modifications to the codebase would inevitably change the behaviour of outputs. By placing in safeguards in terms of automated testing (since it is software that can be run automatically by a microprocessor), bugs introduced during the development can be avoided and flagged to the developer in a manner appropriate. The automated testing of the codebase is part of the continuous integration (CI) software development practice, that we would touch upon below.
+
+In general, testing can be broken into levels of granularity. At the lowest level of granularity, we have unit tests, suitably written for testing the small chunks of code, generally functions and basis properties of structures. Climbing one level up, we have integration testing. Here, the focus is more on validation of correct software behaviour between inputs and outputs of functions. As software written gets complicated, systems level testing is introduced as the next level of granularity, which now is considered a black-box technique, usually tasked to be executed by software quality engineers instead of developers and hence we omit ourselves from the systems testing and higher level granularity at this point.
+
+In terms of unit test written, coverage has been written for `DomainDesriptors.jl`, `States.jl`, `LinearSolvers.jl` and `Transformations.jl`. Due to its nature of being a unit test, this package adopts the convention that the test files are named similarily to their respectively source files, and structure in a mirrored way to `src`. In Julia's `Test.jl` package, developers are able to assign test cases and apply assertions via the macro @test.
+```julia
+@test <write statement here>
+```
+This is the core feature of testing, allowing for assertion to be made to ensure the correctness of the statement in the <> braces. Doing so in `test/sim/DomainDescriptors.jl`, we are able to test the constructor of `DomainDescriptor`, ensuring the features expected from the `DomainDescriptor` struct are adhered to.
+
+Furthermore, due to the multiple assertions being made in a test, the `@testset` macro allows developer to sort sets together, organising the test summary into digestable chunks. When organised wisely where each testset are sorted by common features, each testset can elucidate insight for the starting point of debugging the codebase when bugs are detected from testing.
+
+Going one level up, the testing done by `StateTransformations.jl` starts integrating the properties of `State` which are to be transformed by properties in a `Transformer`. Since the interaction here are due to functions involving a `util` and a `sim` dependecy, it is considered as an integration test, and have resides in the `StateTransformations.jl` test file in `test/`.
```

Comment:
```suggestion
Going one level up, the testing done by `StateTransformations.jl` starts integrating the properties of `State` which are to be transformed by properties in a `Transformer`. Since the interaction here are due to functions involving a `util` and a `sim` dependency, it is considered as an integration test, and have resides in the `StateTransformations.jl` test file in `test/`.
```